Job: Respect ControllerRef

enisoc opened the pull request:
**What this PR does / why we need it**:

This is part of the completion of the [ControllerRef](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md) proposal. It brings Job into full compliance with ControllerRef. See the individual commit messages for details.

**Which issue this PR fixes**:

This ensures that Job does not fight with other controllers over control of Pods.

Ref: #24433

**Special notes for your reviewer**:

**Release note**:

```release-note
Job controller now respects ControllerRef to avoid fighting over Pods.
```
cc @erictune @kubernetes/sig-apps-pr-reviews

Commit:
Job: Add ControllerRef on all created Pods.

Commit:
Job: Set DefaultGarbageCollectionPolicy to OrphanDependents.

Now that Job adds ControllerRef to Pods it creates,
we need to set this default so legacy behavior is maintained.

Commit:
Job: Use ControllerRefManager to adopt/orphan.

Commit:
Job: Use ControllerRef to route watch events.

This is part of the completion of ControllerRef, as described here:

https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md#watches

Commit:
Job: Update Lister documentation for ControllerRef.

The Job Listers still use selectors, because this is the
behavior expected by callers. This clarifies the meaning of the
returned list. Some callers may need to switch to using
GetControllerOf() instead, but that is a separate, case-by-case issue.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/42176)
<!-- Reviewable:end -->

soltysh commented on pkg/controller/job/jobcontroller_test.go:
```diff
@@ -274,6 +275,22 @@ func TestControllerSyncJob(t *testing.T) {
+		}
+		// Make sure the ControllerRefs are correct.
+		for _, controllerRef := range fakePodControl.ControllerRefs {
+			if got, want := controllerRef.APIVersion, "batch/v1"; got != want {
```
Can you verify the entire ControllerRef object, iow. including the Name and UID?

soltysh reviewed (approved):
A nit, but mostly LGTM.

kow3ns commented:
Reviewed 31 of 31 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks broke.

---

*[pkg/controller/job/jobcontroller.go, line 51 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42176#-Ke0b5o3VVM_ffDQ6uux:-Ke0b5o3VVM_ffDQ6uuy:b-rdoodb) ([raw file](https://github.com/kubernetes/kubernetes/blob/38c4f5d95158e9040482d95c9c7558ee9461382c/pkg/controller/job/jobcontroller.go#L51)):*
> ```Go
> 
> // ControllerKind contains the schema.GroupVersionKind for this controller type.
> var ControllerKind = batch.SchemeGroupVersion.WithKind("Job")
> ```

Probably should be package private

---

*[pkg/controller/job/jobcontroller.go, line 149 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42176#-Ke0bHk_nlteKT_tzpyf:-Ke0bHk_nlteKT_tzpyg:b-95lgwe) ([raw file](https://github.com/kubernetes/kubernetes/blob/38c4f5d95158e9040482d95c9c7558ee9461382c/pkg/controller/job/jobcontroller.go#L149)):*
> ```Go
> func (jm *JobController) getPodJobs(pod *v1.Pod) []*batch.Job {
> 	jobs, err := jm.jobLister.GetPodJobs(pod)
> 	if err != nil {
> ```

This is already here, but you have been removing th

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42176)*
<!-- Sent from Reviewable.io -->

Commit:
Job: Always set BlockOwnerDeletion in ControllerRef.

Commit:
Job: Fix CronJob e2e test for async Job deletion.

Now that the default delete option for Job is OrphanDependents,
Job deletion is asynchronous.

enisoc commented:
@k8s-bot non-cri e2e test this

Commit:
Job: Check that ControllerRef UID matches.

Commit:
Job: Check ControllerRef Name and UID in unit test.

enisoc commented:
@k8s-bot cvm gce e2e test this

enisoc commented:
@k8s-bot non-cri e2e test this

enisoc commented:
@soltysh Please take a look. I believe all comments are addressed.

soltysh commented:
/lgtm

soltysh commented:
/approve

soltysh commented:
@ethernetdan @calebamiles is this ok to merge for 1.6 release?

caesarxuchao commented:
/approve

Approve the changes to client.

lavalamp commented:
/approve

(registry changes)

soltysh commented:
@lavalamp since head is open for 1.7 I think we can drop do-not-merge, right?

soltysh commented:
@k8s-bot test this

liggitt commented:
have the appropriate fixes been made to avoid racing the garbage controller readopting a pod on an orphaning deletion of a job?

fejta commented:
/assign @lavalamp 

See https://github.com/kubernetes/kubernetes/pull/42176#issuecomment-289751249

enisoc commented:
I've removed LGTM because I have some changes planned to bring this up to date with the other ControllerRef PRs that moved forward during the 1.6 cycle.

Commit:
Job: Recheck DeletionTimestamp before adoption.

This is necessary to avoid racing with the GC when it orphans
dependents.

Commit:
Job: Add PATCH Pods permission.

This is needed to update ControllerRef during adopt/release.

Commit:
Job: Add e2e test for Pod adopt/release.

An e2e test is currently the only way to ensure we have the correct RBAC
permissions to PATCH Pods.

enisoc commented:
This is ready for final re-review.

I've added 3 commits to fix the equivalent of https://github.com/kubernetes/kubernetes/pull/42938 and https://github.com/kubernetes/kubernetes/pull/43034. I made no changes to the previous commits other than rebasing on master.

janetkuo commented on pkg/controller/job/jobcontroller_test.go:
```diff
@@ -598,7 +599,8 @@ func TestJobPodLookup(t *testing.T) {
 		sharedInformerFactory.Batch().V1().Jobs().Informer().GetIndexer().Add(tc.job)
-		if job := manager.getPodJob(tc.pod); job != nil {
+		if jobs := manager.getPodJobs(tc.pod); len(jobs) > 0 {
+			job := jobs[0]
```
Also make sure `len(jobs)==1`

enisoc replied:
Done.

janetkuo reviewed (commented):
LGTM in general

enisoc commented on pkg/controller/job/jobcontroller.go:
```diff
@@ -167,14 +171,32 @@ func (jm *JobController) addPod(obj interface{}) {
+		}
 		jobKey, err := controller.KeyFunc(job)
 		if err != nil {
-			utilruntime.HandleError(fmt.Errorf("Couldn't get key for job %#v: %v", job, err))
```
In a comment on another ControllerRef PR, the reviewer asked me to remove this type of logging during watch handlers.

janetkuo reviewed (approved):
LGTM
